src/v/crypto/test/crypto_bench.cc

```diff
@@ -0,0 +1,268 @@
+    void update(const std::array<char, Size>& data) {
+        update(data.data(), Size);
+    }
+
+    /**
+     * Return the current output and reset.
+     */
+    std::array<char, DigestSize> reset() {
+        std::array<char, DigestSize> digest;
+        gnutls_hmac_output(_handle, digest.data());
+        return digest;
+    }
+
+private:
+    // silence clang-tidy about _handle being uninitialized
+    // NOLINTNEXTLINE(hicpp-member-init, cppcoreguidelines-pro-type-member-init)
+    hmac(const void* key, size_t size) {
+        int ret = gnutls_hmac_init(&_handle, Algo, key, size);
+        if (unlikely(ret)) {
+            throw std::runtime_error(gnutls_strerror(ret));
+        }
+
+        ret = gnutls_hmac_get_len(Algo);
+        if (unlikely(ret != DigestSize)) {
+            throw std::runtime_error("invalid digest length");
+        }
+    }
+
+    void update(const void* data, size_t size) {
+        int ret = gnutls_hmac(_handle, data, size);
+        if (unlikely(ret)) {
+            throw std::runtime_error(gnutls_strerror(ret));
+        }
+    }
+
+    gnutls_hmac_hd_t _handle;
+};
+
+template<gnutls_digest_algorithm_t Algo, size_t DigestSize>
+class hash {
+public:
+    static constexpr auto digest_size = DigestSize;
+    using digest_type = std::array<char, DigestSize>;
+
+    hash() {
+        int ret = gnutls_hash_init(&_handle, Algo);
+        if (unlikely(ret)) {
+            throw std::runtime_error("hash init failed");
+        }
+
+        ret = gnutls_hash_get_len(Algo);
+        if (unlikely(ret != DigestSize)) {
+            throw std::runtime_error("BOO");
+        }
+    }
+
+    hash(const hash&) = delete;
+    hash& operator=(const hash&) = delete;
+    hash(hash&&) = delete;
+    hash& operator=(hash&&) = delete;
+
+    ~hash() noexcept { gnutls_hash_deinit(_handle, nullptr); }
+
+    void update(std::string_view data) { update(data.data(), data.size()); }
+    void update(bytes_view data) { update(data.data(), data.size()); }
+
+    /**
+     * Return the current output and reset.
+     */
+    digest_type reset() {
+        std::array<char, DigestSize> digest;
+        gnutls_hash_output(_handle, digest.data());
+        return digest;
+    }
+
+private:
+    void update(const void* data, size_t size) {
+        int ret = gnutls_hash(_handle, data, size);
+        if (unlikely(ret)) {
+            throw std::runtime_error("blah update");
+        }
+    }
+
+    gnutls_hash_hd_t _handle;
+};
+
+using hmac_sha256 = hmac<GNUTLS_MAC_SHA256, 32>;
+using hmac_sha512 = hmac<GNUTLS_MAC_SHA512, 64>;
+using hash_sha256 = hash<GNUTLS_DIG_SHA256, 32>;
+using hash_sha512 = hash<GNUTLS_DIG_SHA512, 64>;
+using hash_md5 = hash<GNUTLS_DIG_MD5, 16>;
+
+template<typename F>
+static size_t test_body(size_t msg_len, F n) {
+    auto buffer = random_generators::gen_alphanum_string(msg_len);
+    for (auto i = inner_iters; i--;) {
+        auto s = n(buffer);
+        perf_tests::do_not_optimize(s);
```

Comment:
should the entire call of `n(buffer)` be passed to `do_not_optimize`? Here it seems just its return value is